### PR TITLE
Introduce an object creation timeout to completely de-couple channel creation from channelPool

### PR DESCRIPTION
**Pull request by nizarm**
Introduce an object creation timeout to completely de-couple channel creation from channelPooll
Remove retry during failure
Fix ssl handshake timeout isssue

### CODE REVIEW

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java [hunk 1]
```diff
@@ -816,6 +831,40 @@ public long getTime()
     }
   }
 
+  private class ObjectCreationTimeoutCallback implements Callback<T>
```
Comment (ChaoLinkedIn) [inline r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java]: This looks like a duplicate of TimeoutCallback. Can you just reuse that?

Comment (nizarm) [reply r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java]: I wanted to throw a retry exception since it is connection related. I have refactored TimeoutCallback to accommodate this need.

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java [hunk 2]
```diff
@@ -698,17 +709,24 @@ public void onError(final Throwable e)
                 LOG.error("Encountered error while invoking error waiter callback", ex);
               }
             }
+
+            // Now after cancelling all the pending tasks, lets make sure to back off on the creation
+            _rateLimiter.incrementPeriod();
```
Comment (ssheng) [inline r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java]: Why is this moved down?

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java [hunk 3]
```diff
@@ -646,8 +648,10 @@ public void run(final SimpleCallback callback)
           return;
         }
 
-        _lifecycle.create(new Callback<T>()
-        {
+        // Lets not trust the _lifecycle to timely return a response here.
+        // Embedding the callback inside a timeout callback (ObjectCreationTimeoutCallback)
+        // to force a response withing creationTimeout deadline to reclaim the object slot in the pool
```
[the remark this answers is not in the document]
Comment (nizarm) [reply r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java]: This is taken care in the recent push

File: r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java (hunk 2 above)
Comment (nizarm) [reply r2-core/src/main/java/com/linkedin/r2/transport/http/client/AsyncPoolImpl.java]: call to this method '_rateLimiter.incrementPeriod()' will schedule the next pending task for object creation execution before the current thread get a chance to execute cancelPendingTasks. So by moving this down after cancelPendingTasks - we make sure that the rate limiting tasks only submit new requests.

File: r2-netty/src/main/java/com/linkedin/r2/transport/http/client/common/ChannelPoolLifecycle.java [hunk 4]
```diff
@@ -118,19 +118,24 @@ public void create(final Callback<Channel> channelCallback)
...
-          if(evt == SslHandshakeCompletionEvent.SUCCESS){
+          if(evt == SslHandshakeCompletionEvent.SUCCESS)
+          {
             channelCallback.onSuccess(c);
             c.pipeline().remove(CHANNELPOOL_SSL_CALLBACK_HANDLER);
           }
+          else if (evt instanceof SslHandshakeCompletionEvent)
+          {
+            Throwable sslException = ((SslHandshakeCompletionEvent) evt).cause();
+            onError(channelCallback, sslException);
```
Comment (ssheng) [inline r2-netty/src/main/java/com/linkedin/r2/transport/http/client/common/ChannelPoolLifecycle.java]: Do we have guarantee that cause() is not null?

Comment (nizarm) [reply r2-netty/src/main/java/com/linkedin/r2/transport/http/client/common/ChannelPoolLifecycle.java]: Yes if it is not instance of SslHandshakeCompletionEvent.SUCCESS - it is guaranteed to have cause() in it.

https://github.com/netty/netty/blob/00afb19d7a37de21b35ce4f6cb3fa7f74809f2ab/handler/src/main/java/io/netty/handler/ssl/SslHandshakeCompletionEvent.java

But yea - it make sense to not depend on this implementation detail and I will add a null check.

Comment (FreCap) [reply r2-netty/src/main/java/com/linkedin/r2/transport/http/client/common/ChannelPoolLifecycle.java]: Let's add a comment here then

Comment (FreCap) [reply r2-netty/src/main/java/com/linkedin/r2/transport/http/client/common/ChannelPoolLifecycle.java]: Shouldn't this also be a retriable exception?

Comment (nizarm) [reply r2-netty/src/main/java/com/linkedin/r2/transport/http/client/common/ChannelPoolLifecycle.java]: Thank you. good point. I will address this in the next release as I already merged the pull request and Identity do not leverage retry exception.

Comment (nizarm) [reply r2-netty/src/main/java/com/linkedin/r2/transport/http/client/common/ChannelPoolLifecycle.java]: @FreCap  - please let me know if you have any other concerns !
